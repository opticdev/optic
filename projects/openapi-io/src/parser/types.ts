export type ExternalRefHandler = {
  order: number;
  canRead: (file: { url: string }) => boolean;
  read: (file: { url: string }) => Promise<string>;
};
