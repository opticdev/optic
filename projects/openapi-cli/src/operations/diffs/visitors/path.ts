import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { PathComponents, PathComponent } from '../..';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
