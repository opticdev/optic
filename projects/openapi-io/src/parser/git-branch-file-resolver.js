'use strict';
const fs = require('fs');
const { ono } = require('@jsdevtools/ono');
const url = require('@apidevtools/json-schema-ref-parser/lib/util/url');
const {
  ResolverError,
} = require('@apidevtools/json-schema-ref-parser/lib/util/errors');
const exec = require('child_process').exec;
const path = require('path');

module.exports = (gitBaseRepo, branch) => ({
  /**
   * The order that this resolver will run, in relation to other resolvers.
   *
   * @type {number}
   */
  order: 1,

  /**
   * Determines whether this resolver can read a given file reference.
   * Resolvers that return true will be tried, in order, until one successfully resolves the file.
   * Resolvers that return false will not be given a chance to resolve the file.
   *
   * @param {object} file           - An object containing information about the referenced file
   * @param {string} file.url       - The full URL of the referenced file
   * @param {string} file.extension - The lowercased file extension (e.g. ".txt", ".html", etc.)
   * @returns {boolean}
   */
  canRead(file) {
    return url.isFileSystemPath(file.url);
  },

  /**
   * Reads the given file and returns its raw contents as a Buffer.
   *
   * @param {object} file           - An object containing information about the referenced file
   * @param {string} file.url       - The full URL of the referenced file
   * @param {string} file.extension - The lowercased file extension (e.g. ".txt", ".html", etc.)
   * @returns {Promise<Buffer>}
   */
  read(file) {
    return new Promise((resolve, reject) => {
      const toGit = path.relative(gitBaseRepo, file.url);
      const command = `git show ${branch}:${toGit}`;
      try {
        exec(
          command,
          { cwd: gitBaseRepo, maxBuffer: 1024 * 1024 * 100 },
          (err, stdout, stderr) => {
            if (err)
              reject(
                new ResolverError(
                  ono(err.message, `Error opening file "${path}"`),
                  path
                )
              );
            if (stdout) resolve(stdout);
          }
        );
      } catch (err) {
        reject(
          new ResolverError(ono(err, `Error opening file "${path}"`), path)
        );
      }
    });
  },
});
