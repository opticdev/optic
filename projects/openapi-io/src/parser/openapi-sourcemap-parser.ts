import $RefParser from '@apidevtools/json-schema-ref-parser';
// @ts-ignore
import * as $RefParserOptions from '@apidevtools/json-schema-ref-parser/lib/options';
// @ts-ignore
import { dereference } from './insourced-dereference';
import path from 'path';

import fetch from 'node-fetch';
import isUrl from 'is-url';

import { JsonSchemaSourcemap } from './sourcemap';
import { gitBranchResolver } from './resolvers/git-branch-file-resolver';
import { ExternalRefHandler } from './types';
import Bottleneck from 'bottleneck';
// @ts-ignore
import * as customYaml from './insourced-yaml';
import {
  UserDefinedHeadersByUrlPrefix,
  customHttpResolver,
  parseHeadersConfig,
  DEFAULT_HEADERS,
  getMostRelevantHeader,
} from './resolvers/custom-http-ref-handler';
import {
  FlatOpenAPIV2,
  FlatOpenAPIV3,
  FlatOpenAPIV3_1,
} from '@useoptic/openapi-utilities';

export {
  JSONParserError,
  ResolverError,
} from '@apidevtools/json-schema-ref-parser';

export type ParseOpenAPIResult<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
> = {
  jsonLike: T;
  sourcemap: JsonSchemaSourcemap;
};

export async function dereferenceOpenApi<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
>(
  path: string,
  options: {
    externalRefHandler?: ExternalRefHandler;
    externalRefHeaders?: UserDefinedHeadersByUrlPrefix;
  } = {}
): Promise<ParseOpenAPIResult<T>> {
  const resolver = new $RefParser();
  const headersMap = parseHeadersConfig(options.externalRefHeaders ?? []);

  const sourcemap = new JsonSchemaSourcemap(path);
  const resolve = {
    file: options.externalRefHandler,
    customRefHandler: options.externalRefHandler,
    customHttpRefHandler: customHttpResolver(headersMap),
    external: true,
  };
  // Resolve all references
  const resolverResults: $RefParser.$Refs = await resolver.resolve(path, {
    resolve,
    parse: {
      yaml: customYaml,
    },
  });

  const limiter = new Bottleneck({
    maxConcurrent: 20,
  });

  // parse all asts and add to sourcemap
  const cachedUrls = new Set<string>([]);

  await Promise.all(
    resolverResults.paths().map((filePath, index) =>
      limiter.schedule(async () => {
        if (isUrl(filePath)) {
          const inCache = cachedUrls.has(filePath);
          if (!inCache) {
            const headers = {
              ...DEFAULT_HEADERS,
              ...getMostRelevantHeader(filePath, headersMap),
            };
            const response = await fetch(filePath, {
              headers,
            });
            const contents = await response.text();
            cachedUrls.add(filePath);
            sourcemap.addFileIfMissingFromContents(filePath, contents, index);
          }
        } else if (options.externalRefHandler) {
          const contents = await options.externalRefHandler.read({
            url: filePath,
          });
          sourcemap.addFileIfMissingFromContents(filePath, contents, index);
        } else {
          await sourcemap.addFileIfMissing(filePath, index);
        }
      })
    )
  );

  // Dereference all references
  dereference(
    resolver,
    {
      ...$RefParserOptions.defaults,
      path: path,
      dereference: { circular: 'ignore' },
      resolve,
      parse: {
        yaml: customYaml,
      },
    },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap };
}

export async function parseOpenAPIWithSourcemap<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
>(
  path: string,
  options: {
    externalRefHeaders?: UserDefinedHeadersByUrlPrefix;
  } = {}
): Promise<ParseOpenAPIResult<T>> {
  return dereferenceOpenApi(path, {
    externalRefHeaders: options.externalRefHeaders,
  });
}

export async function parseOpenAPIFromRepoWithSourcemap<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
>(
  name: string,
  repoPath: string,
  branch: string,
  options: {
    externalRefHeaders?: UserDefinedHeadersByUrlPrefix;
  } = {}
): Promise<ParseOpenAPIResult<T>> {
  const inGitResolver = gitBranchResolver(repoPath, branch);
  const fileName = path.join(repoPath, name);
  return dereferenceOpenApi(fileName, {
    externalRefHandler: inGitResolver,
    externalRefHeaders: options.externalRefHeaders,
  });
}
