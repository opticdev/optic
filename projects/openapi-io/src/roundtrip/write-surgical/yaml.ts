import {
  ParseResult,
  PatchApplyResult,
  ReduceOperationType,
  RoundtripProvider,
} from '../roundtrip-provider';
import jsonpatch, { AddOperation, Operation } from 'fast-json-patch';
import {
  insertLines,
  lines,
  pad,
  removeLines,
  replaceRange,
  startsWithWhitespace,
  yamlSpacer,
} from '../helpers/lines';
import fs from 'fs-extra';
import {
  Kind,
  safeLoad,
  YamlMap,
  YAMLMapping,
  YAMLSequence,
} from 'yaml-ast-parser';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { resolveJsonPointerInYamlAst } from '../../parser/openapi-sourcemap-parser';
import invariant from 'ts-invariant';
import ast from '../helpers/ast';
import jsonLike from '../helpers/json-like';
import yamlLike from '../helpers/yaml-like';
import { formatYaml } from '../helpers/format/format';
import {
  returnLineForPosition,
  returnLineNumberPosition,
} from '../helpers/next-or-last-char';
import { loadYaml } from '../../write';

const { EOL } = require('os');

export type YamlRoundTripConfig = {
  count: 2 | 4;
};

class YamlRoundtripImpl implements RoundtripProvider<YamlRoundTripConfig> {
  async applyPatches(
    filePath: string,
    fileContents: string,
    operations: Operation[],
    config: YamlRoundTripConfig | undefined
  ): Promise<PatchApplyResult> {
    const writeConfig = config ? config : await this.inferConfig(fileContents);
    const initialDocument = await this.parse(filePath, fileContents);

    invariant(
      initialDocument.success,
      // @ts-ignore
      `could not parse '${filePath}' as json. Error: ${initialDocument.error}`
    );

    // copy because we've discovered patcher may mutate patches when applying
    const ops: Operation[] = JSON.parse(JSON.stringify(operations));

    const reducerInput: ReduceOperationType = {
      contents: fileContents,
      currentValue: initialDocument.value,
    };

    try {
      const updatedDocument = ops.reduce(
        (input: ReduceOperationType, operation: Operation, index: number) => {
          const currentValue = jsonpatch.applyReducer(
            input.currentValue,
            operation,
            index
          );

          const contents = applyJsonPatchOperationToString(
            input.contents,
            currentValue,
            operation,
            writeConfig
          );

          return { currentValue, contents };
        },
        reducerInput
      );

      const formatted = formatYaml(updatedDocument.contents, writeConfig);

      return {
        success: true,
        value: updatedDocument.currentValue,
        asString: formatted,
        filePath,
      };
    } catch (e: any) {
      return { success: false, error: e.message, filePath };
    }
  }

  fileExtensions: string[] = ['.yaml', '.yml'];

  async inferConfig(contents: string): Promise<YamlRoundTripConfig> {
    const asLines = lines(contents);

    // it's all one line, we'll be overwriting today
    if (asLines.length <= 1) {
      return { count: 2 };
    }

    const firstLineWithWhiteSpaceToStart = asLines.find((i) =>
      startsWithWhitespace(i)
    );

    if (firstLineWithWhiteSpaceToStart) {
      const startChar = startsWithWhitespace(
        firstLineWithWhiteSpaceToStart
      ) as string;

      if (startChar.includes('\t')) {
        return {
          count:
            Array.from(startChar).filter((i) => i === '\t').length === 4
              ? 4
              : 2,
        };
      }
      if (startChar.includes(' ')) {
        return {
          count:
            Array.from(startChar).filter((i) => i === '\t').length === 4
              ? 4
              : 2,
        };
      }
    }

    return { count: 2 };
  }

  name: string = 'yaml';

  async parse(filepath: string, contents: string): Promise<ParseResult> {
    try {
      const value = loadYaml(contents);
      return { value, success: true };
    } catch (e: any) {
      return { error: e.message, success: false };
    }
  }
}

export const YamlRoundtripper = new YamlRoundtripImpl();

// Patch Logic
export function applyJsonPatchOperationToString(
  contents: string,
  jsonLikeValue: any,
  operation: Operation,
  writeConfig: YamlRoundTripConfig
): string {
  const astRoot = safeLoad(contents);

  invariant(
    astRoot.errors.length === 0,
    'error loading ast ' + astRoot.errors.length
  );

  switch (operation.op) {
    case 'add': {
      const addOperation = operation as AddOperation<any>;
      const [parentPath, key, childPath] = jsonPointerHelpers.splitParentChild(
        addOperation.path
      );

      const parentAstNode = resolveJsonPointerInYamlAst(astRoot, parentPath);
      const parentJsonLikeNode = jsonPointerHelpers.tryGet(
        jsonLikeValue,
        parentPath
      );

      invariant(
        Boolean(parentAstNode) && parentJsonLikeNode.match,
        `parent AST node at ${parentPath} not found`
      );

      if (parentAstNode && ast.isObject(parentAstNode)) {
        // verify parent type matches AST Type we got
        invariant(
          jsonLike.isObject(parentJsonLikeNode.value),
          'AST node and Json Value must both be object'
        );

        const parentObjectAst = ast.getObjectMapNode(parentAstNode);
        ast.ensureKeyIsFree(parentObjectAst, key);

        const newField = yamlLike.generatedAddedField(
          key,
          operation.value,
          writeConfig
        );

        return addFieldToObject(
          parentObjectAst,
          contents,
          newField,
          writeConfig
        );
      }

      if (parentAstNode && ast.isArray(parentAstNode)) {
        const parentArrayAst = ast.getArrayNode(parentAstNode);

        const newItem = yamlLike.generatedAdded(operation.value, writeConfig);
        const addIndex: number =
          key === '-' ? parentArrayAst.items.length : Number(key);

        invariant(
          !isNaN(addIndex) && addIndex <= parentArrayAst.items.length,
          'array insert index outside of bounds'
        );

        // verify parent type matches AST Type we got
        invariant(
          jsonLike.isArray(parentJsonLikeNode.value),
          'AST node and Json Value must both be object'
        );

        return addItemToArray(
          parentArrayAst,
          contents,
          newItem,
          addIndex,
          writeConfig
        );
      }

      invariant(
        false,
        'add operations can only be applied to parents that are ararys or objects'
      );

      break;
    }
    case 'remove': {
      const [parentPath, key, childPath] = jsonPointerHelpers.splitParentChild(
        operation.path
      );

      const parentAstNode = resolveJsonPointerInYamlAst(astRoot, parentPath);
      const parentJsonLikeNode = jsonPointerHelpers.tryGet(
        jsonLikeValue,
        parentPath
      );

      invariant(
        Boolean(parentAstNode) && parentJsonLikeNode.match,
        `parent AST node at ${parentPath} not found`
      );

      if (parentAstNode && ast.isObject(parentAstNode)) {
        const parentObjectAst = ast.getObjectMapNode(parentAstNode);

        // verify parent type matches AST Type we got
        invariant(
          jsonLike.isObject(parentJsonLikeNode.value),
          'AST node and Json Value must both be object'
        );

        invariant(
          Boolean(parentObjectAst.mappings.find((i) => i.key.value === key)),
          `AST mapping node must contain key '${key}' to remove it`
        );

        return removeFieldFromObject(parentObjectAst, contents, key);
      }

      if (parentAstNode && ast.isArray(parentAstNode)) {
        const parentArrayAst = ast.getArrayNode(parentAstNode);
        // verify parent type matches AST Type we got
        invariant(
          jsonLike.isArray(parentJsonLikeNode.value),
          'AST node and Json Value must both be object'
        );

        invariant(
          !isNaN(Number(key)),
          'AST node remove index must be a number: ' + key
        );

        return removeItemFromArray(parentArrayAst, contents, Number(key));
      }

      invariant(
        false,
        'remove operations can only be applied to parents that are ararys or objects'
      );

      break;
    }
    case 'replace': {
      const astNode = resolveJsonPointerInYamlAst(astRoot, operation.path)!;
      invariant(
        astNode,
        'node to update can not be found at ' + operation.path
      );
      const generated = yamlLike.generatedAdded(operation.value, writeConfig);

      const line = returnLineForPosition(contents, astNode.startPosition);
      const leadingWhitespace = startsWithWhitespace(line) || '';

      if (astNode.kind === Kind.MAPPING) {
        const asMapping = astNode as YAMLMapping;
        const replacingWithCollection =
          jsonLike.isObject(operation.value) ||
          jsonLike.isArray(operation.value);
        const leadingEol =
          lines(generated).length > 1 || replacingWithCollection ? EOL : '';
        const replace = pad(
          `${leadingEol}${generated}`,
          leadingWhitespace + yamlSpacer(writeConfig)
        )
          .withLeading(`${leadingWhitespace}${asMapping.key.value}: `)
          .flush();

        return replaceRange(
          contents,
          asMapping.startPosition,
          asMapping.endPosition,
          replace
        );
      } else {
        const spacer = yamlSpacer(writeConfig);
        const replace = pad(generated, leadingWhitespace + spacer)
          .withLeading('')
          .flush();
        return replaceRange(
          contents,
          astNode.startPosition,
          astNode.endPosition,
          replace
        );
      }
    }

    default:
      invariant(
        false,
        'roundtrip json does not support json path: ' + operation.op
      );
  }
}

function addFieldToObject(
  parentAst: YamlMap,
  contents: string,
  fieldString: string,
  writeConfig: YamlRoundTripConfig
) {
  const isEmptyObject = ast.keys(parentAst).length === 0;
  if (isEmptyObject) {
    const line = returnLineForPosition(contents, parentAst.startPosition);
    const leadingWhitespace = startsWithWhitespace(line) || '';
    const spacer = yamlSpacer(writeConfig);
    const replace = pad(fieldString, leadingWhitespace + spacer + spacer)
      .withLeading('\n' + leadingWhitespace + spacer)
      .flush();
    return replaceRange(
      contents,
      parentAst.startPosition,
      parentAst.endPosition,
      replace
    );
  } else {
    const lastField = parentAst.mappings[parentAst.mappings.length - 1];

    const line = returnLineForPosition(contents, lastField.startPosition);
    const leadingWhitespace = startsWithWhitespace(line) || '';

    const insert = pad(fieldString, leadingWhitespace)
      .withLeading('\n' + leadingWhitespace)
      .flush();

    return replaceRange(
      contents,
      lastField.endPosition,
      lastField.endPosition,
      insert
    );
  }
}

function removeFieldFromObject(
  parentAst: YamlMap,
  contents: string,
  key: string
) {
  const keyIndex = parentAst.mappings.findIndex((i) => i.key.value === key);
  const numberOfKeys = parentAst.mappings.length;

  const isOnlyKey = numberOfKeys === 1;

  const item = parentAst.mappings.find((i) => i.key.value === key)!;

  if (isOnlyKey) {
    return replaceRange(
      contents,
      parentAst.startPosition,
      parentAst.endPosition,
      '{}'
    );
  }

  const startLine = returnLineNumberPosition(contents, item.startPosition);

  const endLine = returnLineNumberPosition(contents, item.endPosition);

  return removeLines(contents, startLine, endLine);
}

function removeItemFromArray(
  parentAst: YAMLSequence,
  contents: string,
  index: number
) {
  const toRemove = parentAst.items[index];
  const numberOfKeys = parentAst.items.length;

  const isOnlyItem = numberOfKeys === 1;

  if (isOnlyItem) {
    return replaceRange(
      contents,
      parentAst.startPosition,
      parentAst.endPosition,
      '[]'
    );
  }

  const startLine = returnLineNumberPosition(contents, toRemove.startPosition);

  const endLine = returnLineNumberPosition(contents, toRemove.endPosition);

  return removeLines(contents, startLine, endLine);
}

function addItemToArray(
  parentAst: YAMLSequence,
  contents: string,
  newItem: string,
  index: number,
  writeConfig: YamlRoundTripConfig
) {
  const isAddingLastItem = parentAst.items.length === index;
  const isAddingFirstItem = index === 0;
  const isAddingToEmptyArray =
    isAddingFirstItem && parentAst.items.length === 0;

  if (isAddingToEmptyArray) {
    const line = returnLineForPosition(contents, parentAst.startPosition);
    const parentLine = returnLineNumberPosition(
      contents,
      parentAst.startPosition
    );

    const leadingWhitespace = startsWithWhitespace(line) || '';

    const insert = pad(newItem, leadingWhitespace + '  ')
      .withLeading(`${EOL}${leadingWhitespace}- `)
      .withTrailing(EOL)
      .flush();

    return replaceRange(
      contents,
      parentAst.startPosition,
      parentAst.endPosition,
      insert
    );
  } else if (isAddingFirstItem) {
    // not done yet
    const itemAfter = parentAst.items[index + 1];
    const line = returnLineForPosition(contents, itemAfter.startPosition);
    const afterLine = returnLineNumberPosition(
      contents,
      itemAfter.startPosition
    );

    const leadingWhitespace = startsWithWhitespace(line) || '';

    const insert = pad(newItem, leadingWhitespace + '  ')
      .withLeading(`${leadingWhitespace}- `)
      .flush();

    return insertLines(contents, afterLine - 1, insert);
  } else if (isAddingLastItem) {
    const itemBefore = parentAst.items[index - 1];
    const line = returnLineForPosition(contents, itemBefore.startPosition);
    const leadingWhitespace = startsWithWhitespace(line) || '';

    const insert = pad(newItem, leadingWhitespace + '  ')
      .withLeading(EOL + leadingWhitespace + '- ')
      .flush();

    return replaceRange(
      contents,
      parentAst.endPosition - 1,
      parentAst.endPosition - 1,
      insert
    );
  } else {
    const itemBefore = parentAst.items[index - 1];
    const line = returnLineForPosition(contents, itemBefore.startPosition);

    const beforeLine = returnLineNumberPosition(
      contents,
      itemBefore.startPosition
    );

    const leadingWhitespace = startsWithWhitespace(line) || '';

    const insert = pad(newItem, leadingWhitespace + '  ')
      .withLeading(leadingWhitespace + '- ')
      .flush();

    return insertLines(contents, beforeLine + 1, insert);
  }
}
