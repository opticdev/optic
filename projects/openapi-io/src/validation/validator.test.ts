import yaml from 'yaml';
import { test, expect, describe } from '@jest/globals';
import fs from 'node:fs/promises';

import {
  validateSwaggerV2Document,
  validateOpenApiV3Document,
} from './validator';
import { parseOpenAPIWithSourcemap } from '../parser/openapi-sourcemap-parser';
import path from 'path';
import { defaultEmptySpec } from '@useoptic/openapi-utilities';

async function readJson(p: string) {
  const contents = await fs.readFile(p, 'utf-8');
  return JSON.parse(contents);
}

describe('2.x.x validation', () => {
  describe('strict validation', () => {
    test('valid swagger document should not raise errors', async () => {
      const file = yaml.parse(
        await fs.readFile('./inputs/swagger2/spec.yml', 'utf-8')
      );

      validateSwaggerV2Document(file);
    });

    test('invalid swagger document should raise errors', async () => {
      expect(() => {
        validateSwaggerV2Document({
          swagger: '2.0',
          info: {
            title: '123',
            version: '1.2.3',
          },
          bad: '123',
          paths: {
            '/api': {
              get: {
                responses: {},
              },
            },
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('non-strict validation', () => {
    test('valid swagger document should not raise errors', async () => {
      const file = yaml.parse(
        await fs.readFile('./inputs/swagger2/spec.yml', 'utf-8')
      );

      validateSwaggerV2Document(file);
    });

    test('invalid swagger document should raise errors', async () => {
      expect(() => {
        validateSwaggerV2Document({
          swagger: '1',
          info: {
            title: '123',
          },
          paths: {
            '/api': {
              get: {},
            },
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('3.x.x validation', () => {
  describe('strict validation', () => {
    test('valid open api document should not raise errors', async () => {
      validateOpenApiV3Document(defaultEmptySpec);
      validateOpenApiV3Document(
        (await readJson('./inputs/openapi3/petstore0.json.flattened.json'))
          .jsonLike
      );
    });

    test('valid open 3.1 api document should not raise errors', async () => {
      validateOpenApiV3Document(defaultEmptySpec);
      validateOpenApiV3Document(
        await readJson('./inputs/openapi3/todo-api-3_1.json')
      );
    });

    test('open api doc with no description in response should throw', () => {
      expect(() => {
        validateOpenApiV3Document(
          {
            openapi: '3.1.3',
            info: { version: '0.0.0', title: 'Empty' },
            paths: {
              '/example': {
                get: {
                  responses: {
                    '200': {},
                  },
                },
              },
            },
          },
          undefined,
          { strictOpenAPI: true }
        );
      }).toThrowErrorMatchingSnapshot();
    });

    test('advanced validators run and append their results', () => {
      const json: any = {
        ...defaultEmptySpec,
        paths: {
          '/api/users/{userId}': {
            get: {
              responses: {
                '200': {
                  description: 'hello',
                  content: {
                    'application/json': {
                      schema: {
                        type: 'object',
                        oneOf: [],
                        anyOf: [],
                        items: [],
                      },
                    },
                  },
                },
              },
            },
          },
        },
      };

      expect(() => {
        validateOpenApiV3Document(json, undefined, { strictOpenAPI: true });
      }).toThrowErrorMatchingSnapshot();
    });

    test('open api doc with extra custom parameters', () => {
      validateOpenApiV3Document({
        ...defaultEmptySpec,
        'x-extra_property': {
          abc: 'asd',
        },
      });

      validateOpenApiV3Document(
        {
          ...defaultEmptySpec,
          paths: {
            '/user/login': {
              get: {
                tags: ['user'],
                'x-maturity': 'wip',
                summary: 'Logs user into the system',
                operationId: 'loginUser',
                parameters: [
                  {
                    name: 'username',
                    in: 'query',
                    description: 'The user name for login',
                    required: true,
                    schema: {
                      type: 'string',
                    },
                  },
                  {
                    name: 'password',
                    in: 'query',
                    description: 'The password for login in clear text',
                    required: true,
                    schema: {
                      type: 'string',
                    },
                  },
                ],
                responses: {
                  '200': {
                    description: 'successful operation',
                    headers: {
                      'X-Rate-Limit': {
                        description: 'calls per hour allowed by the user',
                        schema: {
                          type: 'integer',
                          format: 'int32',
                        },
                      },
                      'X-Expires-After': {
                        description: 'date in UTC when token expires',
                        schema: {
                          type: 'string',
                          format: 'date-time',
                        },
                      },
                    },
                    content: {
                      'application/xml': {
                        schema: {
                          type: 'string',
                        },
                      },
                      'application/json': {
                        schema: {
                          type: 'string',
                        },
                      },
                    },
                  },
                  '400': {
                    description: 'Invalid username/password supplied',
                    content: {},
                  },
                },
              },
            },
          },
        },
        undefined,
        { strictOpenAPI: true }
      );
    });

    test('openapi with webhooks', async () => {
      validateOpenApiV3Document(
        await readJson('./inputs/openapi3/openapi-webhook.json')
      );
    });
  });

  describe('non-strict validation', () => {
    describe('should pass', () => {
      test('open api doc with no description in response', () => {
        validateOpenApiV3Document(
          {
            openapi: '3.1.3',
            info: { version: '0.0.0', title: 'Empty' },
            paths: {
              '/example': {
                get: {
                  responses: {
                    '200': {},
                  },
                },
              },
            },
          },
          undefined,
          { strictOpenAPI: false }
        );
      });

      test('additional properties in invalid place', () => {
        validateOpenApiV3Document(
          {
            openapi: '3.1.3',
            info: { version: '0.0.0', title: 'Empty', badproperty: ':(' },
            paths: {
              '/example': {
                get: {
                  responses: {
                    '200': {},
                  },
                },
              },
            },
          },
          undefined,
          { strictOpenAPI: false }
        );
      });
    });

    describe('should fail', () => {
      test('open api doc without responses', () => {
        expect(() => {
          validateOpenApiV3Document(
            {
              openapi: '3.1.3',
              info: { version: '0.0.0', title: 'Empty' },
              paths: {
                '/example': {
                  get: {},
                },
              },
            },
            undefined,
            { strictOpenAPI: false }
          );
        }).toThrowErrorMatchingSnapshot();
      });

      test('open api doc with invalid status code shape', () => {
        expect(() => {
          validateOpenApiV3Document(
            {
              openapi: '3.1.3',
              info: { version: '0.0.0', title: 'Empty' },
              paths: {
                '/example': {
                  get: {
                    responses: {
                      '202': null,
                    },
                  },
                },
              },
            },
            undefined,
            { strictOpenAPI: false }
          );
        }).toThrowErrorMatchingSnapshot();
      });

      test('open api doc with no path should throw an error', () => {
        expect(() => {
          validateOpenApiV3Document(
            {
              openapi: '3.1.3',
              info: { version: '0.0.0', title: 'Empty' },
            },
            undefined,
            { strictOpenAPI: false }
          );
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

test('processValidatorErrors', () => {
  const json: any = {
    ...defaultEmptySpec,
    paths: {
      '/api/users/{userId}': {
        get: {
          responses: {
            '200': {
              description: 'hello',
              headers: { isgood: { schema: {} } },
              content: {
                'application/json': {
                  schema: {
                    type: 'object',
                    properties: {
                      hello: {
                        type: 'array',
                        items: 'no',
                      },
                    },
                  },
                },
              },
            },
            '400': {
              description: 'hello',
              headers: { isgood: { schema: {} } },
              content: {
                'application/json': {
                  schema: {},
                },
              },
            },
          },
        },
      },
    },
  };
  expect(() => {
    validateOpenApiV3Document(json, undefined, { strictOpenAPI: true });
  }).toThrowErrorMatchingSnapshot();
});

test('processValidatorErrors attaches the sourcemap', async () => {
  const spec = await parseOpenAPIWithSourcemap(
    path.join(__dirname, '../../inputs/openapi3/broken-open-api.json')
  );

  try {
    validateOpenApiV3Document(spec.jsonLike, spec.sourcemap, {
      strictOpenAPI: true,
    });
  } catch (error) {
    const filterOutFileNames = (error as any).message
      .split('\n')
      .filter((i: any) => !i.startsWith('[90m/'))
      .join('\n');

    expect(filterOutFileNames).toMatchSnapshot();
  }
});
