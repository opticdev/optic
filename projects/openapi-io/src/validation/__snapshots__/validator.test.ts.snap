// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`2.x.x validation non-strict validation invalid swagger document should raise errors 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /api > get must have required property 'responses'[39m[22m
/paths/~1api/get
[31minvalid openapi: [39m[1m[31mswagger must be equal to one of the allowed values 2.0[39m[22m
/swagger
[31minvalid openapi: [39m[1m[31minfo must have required property 'version'[39m[22m
/info"
`;

exports[`2.x.x validation strict validation invalid swagger document should raise errors 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /api > get > responses must NOT be valid[39m[22m
/paths/~1api/get/responses"
`;

exports[`3.x.x validation non-strict validation should fail open api doc with invalid status code shape 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /example > get > responses > 202 must be object[39m[22m
/paths/~1example/get/responses/202"
`;

exports[`3.x.x validation non-strict validation should fail open api doc with no path should throw an error 1`] = `
"[31minvalid openapi: [39m[1m[31m must have required property 'paths'[39m[22m
"
`;

exports[`3.x.x validation non-strict validation should fail open api doc without responses 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /example > get must have required property 'responses'[39m[22m
/paths/~1example/get"
`;

exports[`3.x.x validation strict validation advanced validators run and append their results 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /api/users/{userId} > get > responses > 200 > content > application/json > schema > oneOf must NOT have fewer than 1 items[39m[22m
/paths/~1api~1users~1{userId}/get/responses/200/content/application~1json/schema/oneOf
[31minvalid openapi: [39m[1m[31mpaths > /api/users/{userId} > get > responses > 200 > content > application/json > schema > anyOf must NOT have fewer than 1 items[39m[22m
/paths/~1api~1users~1{userId}/get/responses/200/content/application~1json/schema/anyOf
[31minvalid openapi: [39m[1m[31mpaths > /api/users/{userId} > get > responses > 200 > content > application/json > schema > items must be object[39m[22m
/paths/~1api~1users~1{userId}/get/responses/200/content/application~1json/schema/items
[31minvalid openapi: [39m[1m[31mpaths > /api/users/{userId} > get > responses > 200 > content > application/json > schema schema with type "object" cannot also include keywords: items[39m[22m
/paths/~1api~1users~1{userId}/get/responses/200/content/application~1json/schema"
`;

exports[`3.x.x validation strict validation open api doc with no description in response should throw 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /example > get > responses > 200 must have required property 'description'[39m[22m
/paths/~1example/get/responses/200"
`;

exports[`processValidatorErrors 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /api/users/{userId} > get > responses > 200 > content > application/json > schema > properties > hello > items must be object[39m[22m
/paths/~1api~1users~1{userId}/get/responses/200/content/application~1json/schema/properties/hello/items"
`;

exports[`processValidatorErrors attaches the sourcemap 1`] = `
"[31minvalid openapi: [39m[1m[31mpaths > /api/users/{userId} > get > responses > 200 > content > application/json > schema > properties > hello > items must be object[39m[22m
[90m18 |[39m   "properties": {
[90m19 |[39m     "hello": {
[90m20 |[39m       "type": "array",
[90m21 |[39m [1m[31m      "items": "no"[39m[22m
[90m22 |[39m     }
[90m23 |[39m   }
[90m24 |[39m }
[31minvalid openapi: [39m[1m[31mtype must be equal to one of the allowed values string,number,integer,object,array,boolean[39m[22m
[90m35 |[39m "content": {
[90m36 |[39m   "application/json": {
[90m37 |[39m     "schema": {
[90m38 |[39m [1m[31m      "type": "not-real"[39m[22m
[90m39 |[39m     }
[90m40 |[39m   }
[90m41 |[39m }"
`;
